spec/ruby/optional/capi/module_spec.rb

```diff
@@ -241,26 +241,44 @@ def method_to_be_aliased
   describe "rb_define_method" do
     it "defines a method on a class" do
       cls = Class.new
-      @m.rb_define_method(cls, "test_method")
+      @m.rb_define_method(cls, "test_method", 0)
       cls.should have_instance_method(:test_method)
       cls.new.test_method.should == :test_method
     end
 
-    it "returns the correct arity of the method in class" do
+    it "returns the correct arity when argc of the method in class is 0" do
       cls = Class.new
-      @m.rb_define_method(cls, "test_method")
+      @m.rb_define_method(cls, "test_method", 0)
       cls.new.method(:test_method).arity.should == 0
     end
 
+    it "returns the correct arity when argc of the method in class is -2" do
+      cls = Class.new
+      @m.rb_define_method(cls, "test_method", -2)
+      cls.new.method(:test_method).arity.should == -1
```

Comment:
I don't quite understand how the expected arity for this argc is `-1`, but CRuby seems to think so.

Reply:
It’s because negative values of `rb_define_method`’s `argc` have a different meaning from the negative values of `Method#arity`.

`rb_define_method` supports [three different “modes”](https://github.com/ruby/ruby/blob/master/doc/extension.rdoc#label-Method+and+Singleton+Method+Definition) for how Ruby will call the C function you pass in:

* fixed number of arguments, e.g. `my_c_function(VALUE self, VALUE arg1, VALUE arg2, VALUE arg3)` for a C function which always takes 3 arguments (in addition to `self`);
* variable number of arguments _in a C array_, i.e. `my_c_function(int argc, VALUE *argv, VALUE self)`, where `argv` is the pointer to the first element of a C array containing the arguments and `argc` says how many elements that array has; and
* variable number of arguments _in a Ruby `Array`_, i.e. `my_c_function(VALUE self, VALUE args)`, where `args` is a Ruby `VALUE` which is guaranteed to be an `Array` instance.

So depending on what’s convenient for the C function you’ve written, you can choose which of these three modes you want Ruby to use when it calls into it. The first mode is the default and that’s what you get if you pass in a non-negative `argc` like `0` or `3`.

The negative `rb_define_method` `argc` values are a mode selection hack: instead of supplying an additional argument to specify the mode, or calling a different function (with a different name) for each mode, you pass in `-1` to choose the “variable number of arguments in a C array” mode and `-2` to choose the “variable number of arguments in a Ruby `Array` mode”. But both of these modes define a Ruby method with a variable number of arguments (and zero required arguments), for which the correct `Method#arity` is `-1`.

I’ll let you draw your own conclusions about whether this is a good API design or not. 😄